drivers/overlay/overlay.go

```diff
@@ -162,3 +150,51 @@ func (d *driver) configure() error {
 func (d *driver) Type() string {
 	return networkType
 }
+
+func (d *driver) nodeJoin(node net.IP, self bool) {
+	d.Lock()
+	if self {
+		d.bindAddress = node.String()
+	} else {
+		d.neighIP = node.String()
+	}
+	d.Unlock()
+	if d.bindAddress != "" && !d.isSerfAlive() {
+		err := d.serfInit()
+		if err != nil {
+			logrus.Errorf("initializing serf instance failed: %v", err)
+			return
+		}
+	}
+
+	if d.serfInstance != nil && d.neighIP != "" {
+		var err error
+		d.joinOnce.Do(func() {
+			err = d.serfJoin()
+		})
+		if err != nil {
+			logrus.Errorf("joining serf neighbor %s failed: %v", d.neighIP, err)
+			d.Lock()
+			d.joinOnce = sync.Once{}
+			d.Unlock()
+			return
+		}
+	}
+}
+
+// DiscoverNew is a notification for a new discovery event, such as a new node joining a cluster
+func (d *driver) DiscoverNew(dType driverapi.DiscoveryType, data interface{}) error {
+	if dType == driverapi.NodeDiscovery {
+		nodeData, ok := data.(driverapi.NodeDiscoveryData)
+		if !ok {
+			return fmt.Errorf("invalid discovery data")
+		}
+		d.nodeJoin(net.ParseIP(nodeData.Address), nodeData.Self)
```

Comment:
It seems like  `nodeJoin` does need the node IP address in string from only, in fact converts it back to string.
It looks like you do not need to convert it to `net.IP` here, you'd avoid the double conversion.